YAML file for R Group Replacements

DavidACosgrove opened the pull request:
It defaults to Layer 1 on, Layer 2 off.
It has 2 input fields for the core, but only one should be used at a time.  They should probably be set up as some sort of mutually-exclusive arrangement.
The script will prefer the SMARTS core if both are provided.

jones-gareth commented:
Instead of two input fields (one for smarts and one for the sketcher) smarts input should be handled by the paste entry and the molecule type of smarts (see the Reaction Pair Finder YAML).

I've run this on the AZ Pyrrolamides and CHEMBL2321810 dataset and I failed to find any analogues.  We certainly need to understand what is happening as these are 2 or our standard RGD datasets.

One reason to use the AZ dataset is to see if we should support multiple cores and wildcards in cores.

DavidACosgrove commented:
Are those datasets anywhere handy?   I’m sure I could find the chembl one,
but I’m not familiar with the AZ set.

On Mon, 5 Sep 2022 at 21:42, Gareth Jones ***@***.***> wrote:

> Instead of two input fields (one for smarts and one for the sketcher)
> smarts input should be handled by the paste entry and the molecule type of
> smarts (see the Reaction Pair Finder YAML).
>
> I've run this on the AZ Pyrrolamides and CHEMBL2321810 dataset and I
> failed to find any analogues. We certainly need to understand what is
> happening as these are 2 or our standard RGD datasets.
>
> One reason to use the AZ dataset is to see if we should support multiple
> cores and wildcards in cores.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1237447920>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FTUCJM2O2VUC5MVIPTV4ZLMXANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

jones-gareth commented:
> Are those datasets anywhere handy? I’m sure I could find the chembl one, but I’m not familiar with the AZ set.
> On Mon, 5 Sep 2022 at 21:42, Gareth Jones ***@***.***> wrote: Instead of two input fields (one for smarts and one for the sketcher) smarts input should be handled by the paste entry and the molecule type of smarts (see the Reaction Pair Finder YAML). I've run this on the AZ Pyrrolamides and CHEMBL2321810 dataset and I failed to find any analogues. We certainly need to understand what is happening as these are 2 or our standard RGD datasets. One reason to use the AZ dataset is to see if we should support multiple cores and wildcards in cores. —

For the Chembl one look at https://github.com/PatWalters/Free-Wilson in the data directory.  Interestingly, I don't think you can get it from Chembl anymore.  

The AZ one is at https://www.dropbox.com/s/cawfsf4oxzx84qb/AZ_Pyrrolamides_4OE_noqualifiers_log_rgd.dxp?dl=0  It is a public Gyrase dataset

DavidACosgrove commented:
For the Chembl one, I downloaded the SMILES file, converted it to CSV and
read it in.  I downloaded the relevant scaffold file and loaded it in in
the dialog with the 'File' option.  It ran but gave me no analogues.  I
then used the edit functionality and at that point I could see that there
were R1, R2 and R3 groups on the scaffold.  I removed those, and changed
the resulting OH to O (not sure if that was necessary) and then I did get
analogues.
For the AZ set, I got a good set of analogues with the SMARTS
c1ccccc1C2CCC(CC2)NC(=O)c3cccn3 (I think, I'm not sure how I rediscover
that).  The key thing there was that the dialogue defaulted to using the
'Core' column for input structures.  When I left it at that I got nothing,
not even an empty table.  When I changed it to the correct column 'Molecule
SMILES' it gave me the analogues as expected.

I think there's a difference in expected behaviour here, rather than
technical problems with the code.  Clearly, though, the code needs to
behave as the user expects it to.  Not being an experienced user of
Spotfire/ChemCharts, I guess my expectations are different.  Did you, for
example, expect it to pick up the Cores column and use those?  If so, we
definitely need to deal with the R Group issue, maybe replacing them with
wildcards in the query.  Maybe something to chat about, now we have
something to point at as a working implementation.

On Mon, Sep 5, 2022 at 10:58 PM Gareth Jones ***@***.***>
wrote:

> Are those datasets anywhere handy? I’m sure I could find the chembl one,
> but I’m not familiar with the AZ set.
> On Mon, 5 Sep 2022 at 21:42, Gareth Jones *@*.***> wrote: Instead of two
> input fields (one for smarts and one for the sketcher) smarts input should
> be handled by the paste entry and the molecule type of smarts (see the
> Reaction Pair Finder YAML). I've run this on the AZ Pyrrolamides and
> CHEMBL2321810 dataset and I failed to find any analogues. We certainly need
> to understand what is happening as these are 2 or our standard RGD
> datasets. One reason to use the AZ dataset is to see if we should support
> multiple cores and wildcards in cores. —
>
> For the Chembl one look at https://github.com/PatWalters/Free-Wilson in
> the data directory. Interestingly, I don't think you can get it from Chembl
> anymore.
>
> The AZ one is at
> https://www.dropbox.com/s/cawfsf4oxzx84qb/AZ_Pyrrolamides_4OE_noqualifiers_log_rgd.dxp?dl=0
> It is a public Gyrase dataset
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1237489834>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FXMKXI3OB4RGT7AS2TV4ZUIHANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

jones-gareth commented:
> 

I think the core column as default is fine- it is ordering columns with mol block columns first I think.  It is reasonable to have the user select the right structure column when there are multiple in the table.

I naively thought that I could use the core with the rgroups present.  I followed your recipe of removing the R groups from the sketched core and I got this error

```
Traceback (most recent call last):
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\script_runner.py", line 22, in <module>
    response = df.execute(request)
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 319, in execute
    analogues_table = replace_rgroups(mols, core_query, use_layer1,
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 284, in replace_rgroups
    if mol is not None and mol and mol.HasSubstructMatch(core_query):
Boost.Python.ArgumentError: Python argument types in
    Mol.HasSubstructMatch(Mol, ChemicalReaction)
did not match C++ signature:
    HasSubstructMatch(class RDKit::ROMol self, class RDKit::MolBundle query, struct RDKit::SubstructMatchParameters params=True)
    HasSubstructMatch(class RDKit::ROMol self, class RDKit::ROMol query, struct RDKit::SubstructMatchParameters params)
    HasSubstructMatch(class RDKit::ROMol self, class RDKit::MolBundle query, bool recursionPossible=True, bool useChirality=False, bool useQueryQueryMatches=False)
    HasSubstructMatch(class RDKit::ROMol self, class RDKit::ROMol query, bool recursionPossible=True, bool useChirality=False, bool useQueryQueryMatches=False)
```

Which is odd.  One difference between my branch and yours is that there are a couple of changes to master.  Could you rebase on master and see if you get this error?  I hope we're not running into those boost binding issues.

jones-gareth commented on python/local/RGroupReplacement.yaml:
```diff
@@ -0,0 +1,84 @@
+  request:
+    id: useLayer2
+    dataType: boolean
+    data: !!bool false
```
The sketcher is set for reactions, but we want a substructure query.

DavidACosgrove replied:
I have attempted to fix this.  It now uses a SMARTS query correctly to generate 72 products from the Chembl set, but what I think is the corresponding MOLfile query gives nothing.  I am wondering if this is because it's not being built as a query molecule, though I have manually changed the aromatic bond types to type 4 rather than the kekule form.  I'm not getting the error you saw, but also not getting any hits.  I am at a loss, I'm afraid.  I expect I'm not passing a type in correctly somewhere.  A chat would no doubt solve it quickest.

jones-gareth replied:
Happy to chat, but ATM I'm also stumped.  The only issue I've had with queries in mol blocks is that sanitization may not work for partial ring systems.  I'll play with things and see if I can figure out what is happening.

jones-gareth commented on python/local/RGroupReplacement.yaml:
```diff
@@ -0,0 +1,84 @@
+    contentType: chemical/x-mdl-rxnfile
+tags:
+- color: '#50AF28'
+  text: chemistry
```
Remove calcprop tag

mattstahl00 commented:
I'm getting the following errors when I paste an input MOL structure.  Input is just a single structure in the sketcher, correct?

RGroupReplacement data function

Error ValueError: empty bond indices
Request id eaf9aa72-306e-4839-b54d-c0d64ae1a40c in C:\Users\mstah\AppData\Local\Temp\ChemChartsDataFxns\jobs\eaf9aa72-306e-4839-b54d-c0d64ae1a40c failed.

   at PublisherCore.CPythonExecutor.RunLocalPython(DataFxnRequest request, DataFunctionInvocation invocation, CancellationToken token, PythonManager pythonManager)
   at PublisherCore.CPythonExecutor.Run(DataFxnRequest request, DataFunctionInvocation invocation, CancellationToken token, PythonManager pythonManager)
   at PublisherCore.CPythonExecutor.Run(DataFxnRequest request, DataFunctionInvocation invocation, CancellationToken token)
   at ChemistryService.DataFxnExecutorBase.<ExecuteFunctionCore>d__1.MoveNext()

-----------------------

Python data function error

Click to see details of error output and any traceback

If reporting an error on a supported data function, copy this notification text

Traceback (most recent call last):
  File "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\script_runner.py", line 22, in <module>
    response = df.execute(request)
  File "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 319, in execute
    analogues_table = replace_rgroups(mols, core_query, use_layer1,
  File "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 286, in replace_rgroups
    core_and_rgroups = make_core_and_rgroups(mol, core_query)
  File "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 91, in make_core_and_rgroups
    frag_mol = Chem.FragmentOnBonds(mol_cp, bonds_to_go,
ValueError: empty bond indices

jones-gareth commented:
> Hi, Thanks for looking at it. It feels like my job, though, if it’s not obvious to you that it’s me misunderstanding the interface between spotfire and the data function. Though I have to say I have no idea what to try next. The equivalent SMARTS query, also in the test code, does work ok. Dave

I think the issue is explicit hydrogens in the query.  These need to be converted to constraints on the heavy atom neighbor
```python
        core_query = input_field_to_molecule(request, 'coreSketcher')
        core_query = Chem.MergeQueryHs(core_query)
```
This worked with the query
![image](https://user-images.githubusercontent.com/24233741/189032245-772615d3-8a62-42c9-9b22-6e0af32d8d29.png)
However, I'm not sure I understand the analogues table output- the parent is always the same as the analogue.

DavidACosgrove commented:
I think this is now working.  In my case, the problem was that when I
deleted the R group from the template, it turned it into a methyl rather
than deleting it completely, so I had an incorrect query.  Once I'd spotted
that, it worked ok, I think.  I don't think there's a need for
Chem.MergeQueryHs which does something subtly different.  Knocking the
hydrogens off the query at the sketch stage can't do any harm.  From a user
perspective, I would find it very convenient if, after loading a file, I
could see it in the sketcher.  It might be the case that one would want to
read in a MOL file and then edit before running.  That was certainly the
case here, where I loaded the scaffold MOL file with the extra methyl, but
to edit it I had to run the data function, then edit it, when the structure
does turn up in the sketcher, and run it again.

It didn't always have the parent the same as the analogue in the output
table, but when the parent didn't produce an analogue, it was copied into
the table, so it might have looked like it.  I have now removed that, so
only genuine analogues appear in the table.  It does mean that some parents
ie. structures from the input column won't feature in the table at all.  I
suppose there could be a blank analogue cell in that case, to make the
point.

On Thu, Sep 8, 2022 at 5:10 AM Gareth Jones ***@***.***>
wrote:

> Hi, Thanks for looking at it. It feels like my job, though, if it’s not
> obvious to you that it’s me misunderstanding the interface between spotfire
> and the data function. Though I have to say I have no idea what to try
> next. The equivalent SMARTS query, also in the test code, does work ok. Dave
>
> I think the issue is explicit hydrogens in the query. These need to be
> converted to constraints on the heavy atom neighbor
>
>         core_query = input_field_to_molecule(request, 'coreSketcher')
>
>         core_query = Chem.MergeQueryHs(core_query)
>
> This worked with the query
> [image: image]
> <https://user-images.githubusercontent.com/24233741/189032245-772615d3-8a62-42c9-9b22-6e0af32d8d29.png>
> However, I'm not sure I understand the analogues table output- the parent
> is always the same as the analogue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1240193573>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FSTQPFKQNLI6HZ2ALTV5FRKDANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

jones-gareth commented:
> I think this is now working. In my case, the problem was that when I deleted the R group from the template, it turned it into a methyl rather than deleting it completely, so I had an incorrect query. Once I'd spotted that, it worked ok, I think. I don't think there's a need for Chem.MergeQueryHs which does something subtly different. Knocking the hydrogens off the query at the sketch stage can't do any harm. From a user perspective, I would find it very convenient if, after loading a file, I could see it in the sketcher. It might be the case that one would want to read in a MOL file and then edit before running. That was certainly the case here, where I loaded the scaffold MOL file with the extra methyl, but to edit it I had to run the data function, then edit it, when the structure does turn up in the sketcher, and run it again. It didn't always have the parent the same as the analogue in the output table, but when the parent didn't produce an analogue, it was copied into the table, so it might have looked like it. I have now removed that, so only genuine analogues appear in the table. It does mean that some parents ie. structures from the input column won't feature in the table at all. I suppose there could be a blank analogue cell in that case, to make the point.
> […](#)

- In my test I had 2 explicit hydrogens in the core so I did need to do the mergequeryhs or the core would never match any of the structures.  I can see use cases where a user might wish to restrict analogs by blocking positions using explicit hydrogens.
- I can talk to Paul to see if the sketcher can be initialized from the paste contents.  I'm not sure how easy it would be to connect those components.
- You can paste a mol block or smiles directly into the sketcher (instead of the paste tab).  If you want to edit the structure this is the best way.  Unfortunately, the ChemDraw sketcher is constantly auto-correcting and inserts explicit hydrogens onto hetero atoms with free valance.  This is extremely frustrating when using the sketcher to enter substructures and cores.  We're talking to PE about a workaround.
- You can also just have the sketcher as a UI component and not have the paste and file tabs (see r group decompositon).  In this case the control is of type sketcher instead of molecule.  This might be better here.
- On closer inspection I do see that there are some analogues.  I think the idea of having an empty cell for a matching structure that has no analogue is a good one.  It's simple for Spotfire to filter the empty cells and get the other view.

mattstahl00 commented:
A few suggested improvements:

1.  Have the user select an 'ID' column and carry the data along to the output.  Let's make it easy for folks to search by ID to see analogs of specific compounds.
2. I'm still seeing compounds in the 'Analogues' column that are identical to the parent.  Unmodified structures should be removed.
3. Any 'Analogues' that appear in the input data set should be removed.  Ignore this if it's already being done.
4. It would be easier for folks to understand the results if the input structure is used as a depiction template.  
5. It would be more useful if the 'replacement' part of the input structures were highlighted, instead of the 'core'.  Or we could use 2 different colors -- blue for the core and red for the replacement.  Let's draw the eye to the bit that changed.
6. It's easy to screw up the core defn by including atoms that are part of an aromatic ring in the core.  Does it make sense to automatically alter the core by deleting aromatic acyclics and re-matching?  If not, we could at least return an error with a human readable explanation.  Something along the lines of "Core definition error.  Please delete core atoms that match only part of a ring in the input structures, or include the entire ring in the core".

Here's the associated error:

RGroupReplacement data function

Error rdkit.Chem.rdchem.AtomKekulizeException: non-ring atom 0 marked aromatic
Request id 400abcd6-2b1e-4a26-b8d5-a9133f3d40b7 in C:\Users\mstah\AppData\Local\Temp\ChemChartsDataFxns\jobs\400abcd6-2b1e-4a26-b8d5-a9133f3d40b7 failed.

   at PublisherCore.CPythonExecutor.RunLocalPython(DataFxnRequest request, DataFunctionInvocation invocation, CancellationToken token, PythonManager pythonManager)
   at PublisherCore.CPythonExecutor.Run(DataFxnRequest request, DataFunctionInvocation invocation, CancellationToken token, PythonManager pythonManager)
   at PublisherCore.CPythonExecutor.Run(DataFxnRequest request, DataFunctionInvocation invocation, CancellationToken token)
   at ChemistryService.DataFxnExecutorBase.<ExecuteFunctionCore>d__1.MoveNext()

DavidACosgrove commented:
My solution to 6 was already lurking in the code, once I had caught and dealt with the sanitization exception that was causing the error you saw.  My example is the c12c(N)cccc2cccc1 with core Nc1ccccc1.  In the initial fragmentation, this produces *cccc* as an R Group needing replacement, and was throwing the AtomKekulizeException.  However, since all the replacements in the table are monodentate, it already puts polydentate fragments back on the core.  So catching the exception, keeping the fragment unsanitized and moving on to the polydentate bit solves the problem.  If there are no other r groups on the core in the offending molecule, no substitutions are found and the molecule is skipped in the output table.

DavidACosgrove commented:
I have added an extra output column in the original table, that gives the number of analogues created.  That way you can easily see which input molecules didn't have work, and modify you core if needed.  Otherwise, you would have to spot missing parents in the output table which is likely to be unreliable.

DavidACosgrove commented:
Could you provide an example of 2, analogues that are identical to parent?  There's a bug if you're seeing them, as I do attempt to remove them.

DavidACosgrove commented:
I've put the code in for the multi-coloured output, with red cores and blue r groups which I now see is the wrong way round.  They different colours aren't being rendered correctly, so that's a moot point at the moment.
It occurs to me that it would be quite simple to colour the layer 1 and layer 2 substituents different colours (red/green was my first thought but colourblind people might object).

jones-gareth commented:
Still need to remove the `calcprop` tag.

jones-gareth commented:
When I sketch using this query 
![image](https://user-images.githubusercontent.com/24233741/191549257-07547737-6812-42ab-98b6-e907cbd33cde.png)
I get no results, though the 2 explicit hydrogens should match all the compounds in the data set.  May need to run `mergeQueryHs` on the query.

jones-gareth commented:
When I run using this query (which I think did work)
![image](https://user-images.githubusercontent.com/24233741/191552232-78290d99-c72a-4ba4-9342-6cd1258e0598.png)
I get this error stack
```python
[09:54:00] SMILES Parse Error: syntax error while parsing: [1[*:0]][[*:0]:1001]
[09:54:00] SMILES Parse Error: Failed parsing SMILES '[1[*:0]][[*:0]:1001]' for input: '[1[*:0]][[*:0]:1001]'
[09:54:00] SMILES Parse Error: syntax error while parsing: [2[*:0]][[*:0]:1003]
[09:54:00] SMILES Parse Error: Failed parsing SMILES '[2[*:0]][[*:0]:1003]' for input: '[2[*:0]][[*:0]:1003]'
[09:54:00] SMILES Parse Error: syntax error while parsing: [3[*:0]][[*:0]:1004]
[09:54:00] SMILES Parse Error: Failed parsing SMILES '[3[*:0]][[*:0]:1004]' for input: '[3[*:0]][[*:0]:1004]'
[09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
[09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
[09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
[09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
Traceback (most recent call last):
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\script_runner.py", line 22, in <module>
    response = df.execute(request)
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 462, in execute
    replace_rgroups(mols, ids, id_column.dataType, core_query, use_layer1,
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 408, in replace_rgroups
    analogues = make_analogues(core_and_rgroups, substs_data,
  File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 283, in make_analogues
    analogue = Chem.CombineMols(analogue, s)
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.CombineMols(Mol, NoneType)
did not match C++ signature:
    CombineMols(class RDKit::ROMol mol1, class RDKit::ROMol mol2, class RDGeom::Point3D offset=<rdkit.Geometry.rdGeometry.Point3D object at 0x00000249A7444140>)


```

mattstahl00 commented:
Dave, here's a multi-core RGD to use for testing the proposed changes -- https://www.dropbox.com/s/mv477msmkutmnkn/GyraseMultiCoreDecomp.dxp?dl=0

Gareth, I proposed that Dave change the DataFxn to take the same inputs as FW -- 'core' and 'rgroups'.  There's no need for the sketcher as the core is already defined.  The DataFxn will 'molzip' cores and R groups, using replacement R groups (where appropriate) in the process.

DavidACosgrove commented:
It may well be, however, that the bug Gareth has found will carry over. It
looks like a hydrogen atom being used as an r group which is likely to be
trouble and in any case doesn’t have an entry in the replacements table.
I’ll take a look tomorrow.

Dave

On Wed, 21 Sep 2022 at 17:57, Matthew Stahl ***@***.***>
wrote:

> Dave, here's a multi-core RGD to use for testing the proposed changes --
> https://www.dropbox.com/s/mv477msmkutmnkn/GyraseMultiCoreDecomp.dxp?dl=0
>
> Gareth, I proposed that Dave change the DataFxn to take the same inputs as
> FW -- 'core' and 'rgroups'. There's no need for the sketcher as the core is
> already defined. The DataFxn will 'molzip' cores and R groups, using
> replacement R groups (where appropriate) in the process.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1253981710>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUGYJZ7PQCYHBAYDNLV7M5ADANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

DavidACosgrove commented:
The test set has rather reinforced my prejudices about automated RGD and
MCS generally.  The cores in this case are too small to be useful for this,
I think, apart from as a pretty brutal test set for missing/unusable data
in cells.  That isn't without large value, of course.  I had to scroll down
to row 331 before I saw an R Group where I thought, "there's one that will
be in the replacements table".  Up to then, the R Groups were either H or
too large and complicated to have possible replacements.  I assume that in
Spotfire you can produce a sub-table based on substructure search?  I think
that's going to be an important teaching point for the final users - in a
large, real-life project set, you will want to run this on tighter
sub-series.  I will concoct some simpler test sets for now.

One potential complication with a large molecule and a small core is the
possibility of the core appearing twice in the molecule, for example a core
of Nc1ccccc1 and analogues Nc1cccc(c1)Nc1ccccc1.  How does the RGD deal
with that?  As a first attempt, I will take the core and R Groups as a set
and use the core match that produces a decomposition that matches the R
Groups also in the table.  Does that sound sensible?

Dave


On Wed, Sep 21, 2022 at 5:57 PM Matthew Stahl ***@***.***>
wrote:

> Dave, here's a multi-core RGD to use for testing the proposed changes --
> https://www.dropbox.com/s/mv477msmkutmnkn/GyraseMultiCoreDecomp.dxp?dl=0
>
> Gareth, I proposed that Dave change the DataFxn to take the same inputs as
> FW -- 'core' and 'rgroups'. There's no need for the sketcher as the core is
> already defined. The DataFxn will 'molzip' cores and R groups, using
> replacement R groups (where appropriate) in the process.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1253981710>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUGYJZ7PQCYHBAYDNLV7M5ADANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

DavidACosgrove commented:
Hi Gareth,
That query works with my CHEMBL2321810 dataset.  Can you send your input
set as well, please?

On Wed, Sep 21, 2022 at 4:56 PM Gareth Jones ***@***.***>
wrote:

> When I run using this query (which I think did work)
> [image: image]
> <https://user-images.githubusercontent.com/24233741/191552232-78290d99-c72a-4ba4-9342-6cd1258e0598.png>
> I get this error stack
>
> [09:54:00] SMILES Parse Error: syntax error while parsing: [1[*:0]][[*:0]:1001]
> [09:54:00] SMILES Parse Error: Failed parsing SMILES '[1[*:0]][[*:0]:1001]' for input: '[1[*:0]][[*:0]:1001]'
> [09:54:00] SMILES Parse Error: syntax error while parsing: [2[*:0]][[*:0]:1003]
> [09:54:00] SMILES Parse Error: Failed parsing SMILES '[2[*:0]][[*:0]:1003]' for input: '[2[*:0]][[*:0]:1003]'
> [09:54:00] SMILES Parse Error: syntax error while parsing: [3[*:0]][[*:0]:1004]
> [09:54:00] SMILES Parse Error: Failed parsing SMILES '[3[*:0]][[*:0]:1004]' for input: '[3[*:0]][[*:0]:1004]'
> [09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
> [09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
> [09:54:00] WARNING: not removing hydrogen atom with dummy atom neighbors
> [09:54:00] WARNING: not removing hydrogen atom with dummy atom neighborsTraceback (most recent call last):
>   File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\script_runner.py", line 22, in <module>
>     response = df.execute(request)
>   File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 462, in execute
>     replace_rgroups(mols, ids, id_column.dataType, core_query, use_layer1,
>   File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 408, in replace_rgroups
>     analogues = make_analogues(core_and_rgroups, substs_data,
>   File "C:\Users\jones\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py", line 283, in make_analogues
>     analogue = Chem.CombineMols(analogue, s)Boost.Python.ArgumentError: Python argument types in
>     rdkit.Chem.rdmolops.CombineMols(Mol, NoneType)did not match C++ signature:
>     CombineMols(class RDKit::ROMol mol1, class RDKit::ROMol mol2, class RDGeom::Point3D offset=<rdkit.Geometry.rdGeometry.Point3D object at 0x00000249A7444140>)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1253907049>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FRM7Q2Q2JYBMVLZNOLV7MV3DANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

DavidACosgrove commented:
Gareth and I are both confused by the behaviour both of pasting in a core
Mol and also reading one from file. The former works ok for SMARTS
patterns.  It seems as though it isn’t being correctly converted into a
query. I need to look at a few things tomorrow that occurred to me in the
gym this evening.
Bedtime now.
D


On Wed, 7 Sep 2022 at 21:59, Matthew Stahl ***@***.***> wrote:

> I'm getting the following errors when I paste an input MOL structure.
> Input is just a single structure in the sketcher, correct?
>
> RGroupReplacement data function
>
> Error ValueError: empty bond indices
> Request id eaf9aa72-306e-4839-b54d-c0d64ae1a40c in
> C:\Users\mstah\AppData\Local\Temp\ChemChartsDataFxns\jobs\eaf9aa72-306e-4839-b54d-c0d64ae1a40c
> failed.
>
> at PublisherCore.CPythonExecutor.RunLocalPython(DataFxnRequest request,
> DataFunctionInvocation invocation, CancellationToken token, PythonManager
> pythonManager)
> at PublisherCore.CPythonExecutor.Run(DataFxnRequest request,
> DataFunctionInvocation invocation, CancellationToken token, PythonManager
> pythonManager)
> at PublisherCore.CPythonExecutor.Run(DataFxnRequest request,
> DataFunctionInvocation invocation, CancellationToken token)
> at ChemistryService.DataFxnExecutorBase.d__1.MoveNext()
> ------------------------------
>
> Python data function error
>
> Click to see details of error output and any traceback
>
> If reporting an error on a supported data function, copy this notification
> text
>
> Traceback (most recent call last):
> File
> "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\script_runner.py",
> line 22, in
> response = df.execute(request)
> File
> "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py",
> line 319, in execute
> analogues_table = replace_rgroups(mols, core_query, use_layer1,
> File
> "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py",
> line 286, in replace_rgroups
> core_and_rgroups = make_core_and_rgroups(mol, core_query)
> File
> "C:\Users\mstah\AppData\Local\Temp\ScriptSync\DataFxnPylib\pylib_3.9.7\df\RGroupReplacement.py",
> line 91, in make_core_and_rgroups
> frag_mol = Chem.FragmentOnBonds(mol_cp, bonds_to_go,
> ValueError: empty bond indices
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Glysade/DataFxns/pull/1#issuecomment-1239875090>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FS4CSNM5YGDUQOCLVLV5D63BANCNFSM6AAAAAAQD3CSEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk
